Core/ControlMapper.cpp

```diff
@@ -88,49 +136,225 @@ void ControlMapper::SetPSPAxis(int device, char axis, float value, int stick) {
+		value = 0.0f;
 	}
+	return value;
 }
 
-bool ControlMapper::Key(const KeyInput &key, bool *pauseTrigger) {
-	std::vector<int> pspKeys;
-	KeyMap::InputMappingToPspButton(InputMapping(key.deviceId, key.keyCode), &pspKeys);
+// Can only be called from Key or Axis.
+bool ControlMapper::UpdatePSPState(const InputMapping &changedMapping) {
+	// Instead of taking an input key and finding what it outputs, we loop through the OUTPUTS and
+	// see if the input that corresponds to it has a value. That way we can easily implement all sorts
+	// of crazy input combos if needed.
+
+	int rotations = 0;
+	switch (g_Config.iInternalScreenRotation) {
+	case ROTATION_LOCKED_HORIZONTAL180: rotations = 2; break;
+	case ROTATION_LOCKED_VERTICAL:      rotations = 1; break;
+	case ROTATION_LOCKED_VERTICAL180:   rotations = 3; break;
+	}
+
+	// For the PSP's button inputs, we just go through and put the flags together.
+	uint32_t buttonMask = 0;
+	uint32_t changedButtonMask = 0;
+	for (int i = 0; i < 32; i++) {
+		uint32_t mask = 1 << i;
+		if (!(mask & CTRL_MASK_USER)) {
+			// Not a mappable button bit
+			continue;
+		}
+
+		uint32_t mappingBit = mask;
+		for (int i = 0; i < rotations; i++) {
+			mappingBit = RotatePSPKeyCode(mappingBit);
+		}
+
+		std::vector<InputMapping> inputMappings;
+		if (!KeyMap::InputMappingsFromPspButton(mappingBit, &inputMappings, false))
+			continue;
+
+		// If a mapping could consist of a combo, we could trivially check it here.
+		for (auto &mapping : inputMappings) {
+			// Check if the changed mapping was involved in this PSP key.
+			if (changedMapping == mapping) {
+				changedButtonMask |= mask;
+			}
+
+			auto iter = curInput_.find(mapping);
+			if (iter != curInput_.end() && iter->second > GetDeviceAxisThreshold(iter->first.deviceId)) {
+				buttonMask |= mask;
+			}
+		}
+	}
+
+	// We only request changing the buttons where the mapped input was involved.
+	setAllPSPButtonStates_(buttonMask & changedButtonMask, (~buttonMask) & changedButtonMask);
+
+	// OK, handle all the virtual keys next. For these we need to do deltas here and send events.
+	for (int i = 0; i < VIRTKEY_COUNT; i++) {
+		int vkId = i + VIRTKEY_FIRST;
+		std::vector<InputMapping> inputMappings;
+		if (!KeyMap::InputMappingsFromPspButton(vkId, &inputMappings, false))
+			continue;
+
+		// If a mapping could consist of a combo, we could trivially check it here.
+		// Save the first device ID so we can pass it into onVKeyDown, which in turn needs it for the analog
+		// mapping which gets a little hacky.
+		float threshold = 1.0f;
+		bool touchedByMapping = false;
+		float value = 0.0f;
+		for (auto &mapping : inputMappings) {
+			if (mapping == changedMapping) {
+				touchedByMapping = true;
+			}
+
+			auto iter = curInput_.find(mapping);
+			if (iter != curInput_.end()) {
+				if (mapping.IsAxis()) {
+					threshold = GetDeviceAxisThreshold(iter->first.deviceId);
+					if (IsUnsignedMapping(vkId)) {
```

Comment:
This really seems to cry out to be a separate method.  Especially with the repeated `value += iter->second;` else cases...

-[Unknown]

Reply:
Will do.